src/fu-engine.c

```diff
@@ -3383,6 +3405,10 @@ fu_engine_write_firmware(FuEngine *self,
 	device_progress = fu_device_progress_new(device, progress);
 	g_return_val_if_fail(device_progress != NULL, FALSE);
 
+	/* notify all the other plugins */
+	if (!fu_engine_composite_peek_firmware(self, device, firmware, progress, flags, error))
```

Comment:
As this is a `peek` function, should it really be failable?  IE should one of these notifying plugins be able to cause a failure for any firmware update?  I'm wondering if it should just be void.

Reply:
How about instead passing as part of composite prepare?  This feels like a better time to fail.

Reply:
the FuDevice and FuFirmware would be different for each device in a composite update -- and plugins only expect composite prepare to be called once per update, no once per device in the update